### DISCUSSION TITLE
Binary/Octal format specifier

### DISCUSSION THREAD

**Discussion by anthony-sv** [dotnet/runtime#75769] [General]
I think I'd be cool and easy to integrate to the language a way to specify Binary or Octal format Specifier, since we can with Hexadecimal, from Standard numeric format specifiers B/b and O/o are free so would have a problem, and we currently have easy way to convert to string with Convert.ToString(value, 2); for binary or from "any" base (more like 2,8,10,16) Convert.ToString(Convert.ToInt32(number, fromBase), toBase);, so we have the way to do this but if it's already working for Hex I don't see why limit it to 1 base, well technically 2 because base 10 is default.

**Comment by CyrusNajmabadi** [comment]
> I think I'd be cool and easy to integrate to the language a way to specify Binary or Octal format Specifier

This is already supported for binary.  can use `0b` as the prefix for that purpsoe.

**Reply by anthony-sv** [reply]
Are you sure? Pretty sure 0b will just be treated as extra text added to the string representation of the object. Hence trying to format for example 255 with 0b will just be 255b

**Reply by FaustVX** [reply]
@anthony-sv 
Have you tried it by yourself ?
[sharplab.io](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRII5wJZICaoC2EhwECyANDCAnAHaV4gDUAPgAIBMAjALAAoQRx4BOABQAGYJJ6yAlAG4gA==)

**Reply by anthony-sv** [reply]
@FaustVX that's a binary literal, I'm talking about Standard Numeric Format Specifiers. i.e. I want to be able to do this:
Console.WriteLine($"{255:B}"); // 11111111

**Reply by svick** [reply]
@anthony-sv This repo is specifically about the C# language, so people interpreted your request is for changing how the language handles numbers in different bases.

But you're talking about standard numeric format strings, which is handled by .Net libraries, not by the C# language. Which means the right place for this request is [dotnet/runtime](https://github.com/dotnet/runtime/).

**Comment by tannergooding** [comment]
Do you have a scenario for Octal in particular? One of the few places I can think of it being used is Linux file permissions.

`0o7` and `0O7` both have large readability issues and its something I'd personally rather not see.

Some languages use `q` instead and so `0q7` or `0Q7` would represent the same. Its more "readable", but isn't really "clear" as to it meaning octal if you aren't familiar with the convention.

**Reply by CyrusNajmabadi** [reply]
FWIW, i was around when the original decisions on this were made (though i was not part of hte decision making).  And the general feeling back then was that C-style octal was just far too niche and confusing.  It was especially undesirable that something like `010` be different from `10`.  We could always use a different prefix, but it still feels like octal just isn't useful enough to warrant that.

**Reply by anthony-sv** [reply]
Mostly for implementing some hardware emulation, and some niche stuff, but in my opinion it's better to have the option even if you won't use it, just like for example in C# we have the option to use operator overloading, where as in Java we don't. About the notation, yeah I don't think an o or 0 would be good for readability and double format is not the C# way, But with O/o I don't think it's that confusing compared to 0 alone, I think that why some made something like 0x for Hex and 0o for Octal and 0b for binary, but in C# are 1 letter specifiers so I don't think O is that bad, but could be Q/q while I agree somebody would have to look for it because its's not "natural" I think that's the case of Hex, we use X instead of H, it's more like a convention now, than "clear". For example in cryptography you would want to use base 64, and we don't even have access to it.

**Comment by CyrusNajmabadi** [comment]
Moving to runtime as this was a request about format-specifiers.  Sorry for hte confusion.  We thought you were asking for C# literal support here.

**Comment by danmoseley** [comment]
@anthony-sv Please create an issue in this repo to if you want to propose the format specifier. You could probably use the API proposal issue template although its not really API. I'm guessing it might go to API review.
